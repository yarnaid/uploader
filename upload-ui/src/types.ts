export type FileType = {
  name: string;
  size: number;
  type?: string;
  updated?: Date;
  created?: Date;
  url: string;
};

export type FilterType = {
  name: string;
  values: string[];
};

export const isFilterType = (obj: any): obj is FilterType => {
  return obj && obj.name && obj.values;
};

export type FileMetadataType = {
  name?: string;
  size?: number;
  type?: string;
  url?: string;
  source?: string;
  created?: string;
  updated?: string;
  tags: FilterType[];
};

export const isFileMetadataType = (obj: any): obj is FileMetadataType => {
  return obj && obj.tags;
};

export type UploadFileType = {
  file: File | null;
  metadata: FileMetadataType | null;
};

export const isUploadFileType = (obj: any): obj is UploadFileType => {
  return obj && obj.file && obj.metadata;
};
