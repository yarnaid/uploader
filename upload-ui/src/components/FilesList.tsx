import React, { useEffect } from "react";
import { useDispatch } from "react-redux";
import { FileType } from "../types";
import { loadFilesListRequest } from "../redux/reducers/files";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import {
  createColumnHelper,
  flexRender,
  getCoreRowModel,
  useReactTable,
  ColumnDef,
} from "@tanstack/react-table";

interface FilesListProps {
  files: FileType[];
}

const columnHelper = createColumnHelper<FileType>();

const columns: ColumnDef<FileType, any>[] = [
  columnHelper.accessor((row) => row.name, {
    id: "name",
    cell: (info) => <a href={`/files/${info.getValue()}`}>{info.getValue()}</a>,
    header: () => <span>Name</span>,
    footer: (info) => info.column.id,
  }),
  columnHelper.accessor("updated", {
    header: () => "Updated",
    cell: (info) => info.renderValue(),
    footer: (info) => info.column.id,
  }),
  columnHelper.accessor("created", {
    header: () => "Created",
    cell: (info) => info.renderValue(),
    footer: (info) => info.column.id,
  }),
  columnHelper.accessor("size", {
    header: () => "Size",
    cell: (info) => info.renderValue(),
    footer: (info) => info.column.id,
  }),
];

const FilesList: React.FC<FilesListProps> = ({ files }) => {
  const dispatch = useDispatch();

  const table = useReactTable({
    data: files,
    columns,
    getCoreRowModel: getCoreRowModel(),
  });

  useEffect(() => {
    dispatch(loadFilesListRequest());
  }, [dispatch]);

  return (
    <Container>
      <Row>
        <Col>
          <h1>Files</h1>
          <table className="table">
            <thead>
              {table.getHeaderGroups().map((headerGroup) => (
                <tr key={headerGroup.id}>
                  {headerGroup.headers.map((header) => (
                    <th key={header.id}>
                      {header.isPlaceholder
                        ? null
                        : flexRender(
                            header.column.columnDef.header,
                            header.getContext()
                          )}
                    </th>
                  ))}
                </tr>
              ))}
            </thead>
            <tbody>
              {table.getRowModel().rows.map((row) => (
                <tr key={row.id}>
                  {row.getVisibleCells().map((cell) => (
                    <td key={cell.id}>
                      {flexRender(
                        cell.column.columnDef.cell,
                        cell.getContext()
                      )}
                    </td>
                  ))}
                </tr>
              ))}
            </tbody>
            <tfoot>
              {table.getFooterGroups().map((footerGroup) => (
                <tr key={footerGroup.id}>
                  {footerGroup.headers.map((header) => (
                    <th key={header.id}>
                      {header.isPlaceholder
                        ? null
                        : flexRender(
                            header.column.columnDef.footer,
                            header.getContext()
                          )}
                    </th>
                  ))}
                </tr>
              ))}
            </tfoot>
          </table>
        </Col>
      </Row>
    </Container>
  );
};

export default FilesList;
