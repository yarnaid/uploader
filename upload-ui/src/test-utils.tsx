import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import { render, RenderOptions } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import configureMockStore from "redux-mock-store";
import { configureStore } from "@reduxjs/toolkit";
import rootReducer from "./redux/reducers";
import {
  FileMetadataType,
  FilterType,
  FileType,
  UploadFileType,
} from "./types";
import type { PreloadedState } from "@reduxjs/toolkit";
import type { RootState, AppStore } from "./redux/store";

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

const mockStore = configureMockStore();

const initialState: {
  files: { files: FileType[] };
  filters: FilterType[];
  file: string;
  tags: FilterType[];
  metadata: FileMetadataType;
} = {
  files: { files: [] },
  filters: [],
  file: "",
  tags: [],
  metadata: { tags: [] },
};

export const store: any = mockStore(initialState);

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <Provider store={store}>
      <MemoryRouter>{children}</MemoryRouter>
    </Provider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: rootReducer, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => render(ui, { wrapper: AllTheProviders, ...options });
export * from "@testing-library/react";

export { customRender as render };
