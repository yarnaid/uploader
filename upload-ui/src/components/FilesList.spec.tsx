import React from "react";
import { render } from "../test-utils";
import { screen } from "@testing-library/react";
import FilesList from "./FilesList";
import { FileType } from "../types";

describe("FilesList", () => {
  const files: FileType[] = [
    {
      name: "test",
      size: 1,
      url: "test",
    },
  ];
  it("renders", () => {
    const { asFragment } = render(<FilesList files={files} />);
    expect(screen.getByText(/Name/)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  it("renders a file", () => {
    render(<FilesList files={files} />);
    expect(screen.getByText(/test/)).toBeInTheDocument();
  });
});
