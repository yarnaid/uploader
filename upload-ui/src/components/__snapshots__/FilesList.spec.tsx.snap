// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilesList renders 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <h1>
          Files
        </h1>
        <table
          class="table"
        >
          <thead>
            <tr>
              <th>
                <span>
                  Name
                </span>
              </th>
              <th>
                Updated
              </th>
              <th>
                Created
              </th>
              <th>
                Size
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                <a
                  href="/files/test"
                >
                  test
                </a>
              </td>
              <td />
              <td />
              <td>
                1
              </td>
            </tr>
          </tbody>
          <tfoot>
            <tr>
              <th>
                name
              </th>
              <th>
                updated
              </th>
              <th>
                created
              </th>
              <th>
                size
              </th>
            </tr>
          </tfoot>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
